sdk/elixir/runtime/main.go

```diff
@@ -0,0 +1,177 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/iancoleman/strcase"
+)
+
+const (
+	ModSourceDirPath     = "/src"
+	sdkSrc               = "/sdk"
+	genDir               = "sdk"
+	schemaPath           = "/schema.json"
+	defaultElixirVersion = "1.16.1-erlang-26.2.2-debian-bookworm-20240130-slim"
+)
+
+func New(
+	// +optional
+	sdkSourceDir *Directory,
+) *ElixirSdk {
+	return &ElixirSdk{SDKSourceDir: sdkSourceDir, RequiredPaths: []string{}}
+}
+
+type ElixirSdk struct {
+	SDKSourceDir  *Directory
+	RequiredPaths []string
+}
+
+func (m *ElixirSdk) ModuleRuntime(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, err
+	}
+
+	modName, err := modSource.ModuleName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subPath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mod := normalizeModName(modName)
+	entrypoint := path.Join(ModSourceDirPath, subPath, mod)
+
+	return ctr.
+		WithEntrypoint([]string{
+			"mix", "cmd",
+			"--cd", entrypoint,
+			"mix do deps.get + dagger.invoke",
+		}), nil
+}
+
+func (m *ElixirSdk) Codegen(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*GeneratedCode, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, fmt.Errorf("could not load module config: %v", err)
+	}
+
+	return dag.GeneratedCode(ctr.Directory(ModSourceDirPath)).
+		WithVCSGeneratedPaths([]string{genDir + "/**"}).
+		WithVCSIgnoredPaths([]string{"dagger"}), nil
+}
+
+func (m *ElixirSdk) CodegenBase(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	modName, err := modSource.ModuleName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subPath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mod := normalizeModName(modName)
+
+	codegenDepsCache, codegenBuildCache := mixProjectCaches(dag, "dagger-codegen")
+
+	ctr := m.Base("").
+		WithMountedDirectory(ModSourceDirPath, modSource.ContextDirectory()).
+		WithMountedDirectory(sdkSrc, m.SDKSourceDir).
+		WithMountedCache(codegenPath()+"/deps", codegenDepsCache).
+		WithMountedCache(codegenPath()+"/_build", codegenBuildCache).
+		With(installCodegen).
+		WithNewFile(schemaPath, ContainerWithNewFileOpts{
+			Contents: introspectionJson,
+		}).
+		WithWorkdir(path.Join(ModSourceDirPath, subPath)).
+		WithDirectory(
+			"dagger",
+			m.SDKSourceDir,
+			ContainerWithDirectoryOpts{Exclude: []string{
+				"*.livemd",
+				"*.md",
+				".changes",
+				"dagger_codegen",
+				"runtime",
+				"scripts",
+				"test",
```

Comment:
Are all of these necessary? Consider excluding what you don't need from the build that puts these files in the engine container, and maybe do some of the steps above like a multi-stage build, so do you don't inherit everything. For example, you can have a separate pipeline for the codegen that just returns the generated files in the end, which you then add here. You won't have any other files from installing codegen, not need to change workdir back, etc, that you don't need afterwards.

Reply:
Actually. I just copy the SDK source from `sdk/elixir` to under sub path and exclude unwanted files here. The reason is I wanted to keep SDK up-to-date without worry about dependencies or source code that's not covered by the codegen, such as graphql core, querybuilder, etc. When you generate by `dagger init` you will got 2 directories under `dagger` directory.

```
wingyplus@WINGYMOMO ~/s/g/w/d/potato (main)> ls -lah dagger
total 24K
drwxr-xr-x 4 wingyplus wingyplus 4.0K Apr 27 15:56 ./
drwxr-xr-x 3 wingyplus wingyplus 4.0K Apr 27 15:56 ../
-rw------- 1 wingyplus wingyplus   27 Apr 27 15:56 .gitattributes
-rw------- 1 wingyplus wingyplus    8 Apr 27 15:56 .gitignore
drwxr-xr-x 3 wingyplus wingyplus 4.0K Apr 27 15:55 dagger/
drwxr-xr-x 4 wingyplus wingyplus 4.0K Apr 27 15:56 potato/
```

The `dagger` is the SDK and the `potato` is the library from module name that added `dagger` as a dependencies.

Reply:
Or I should rename directory to `sdk` to avoid confusion?

Reply:
Yeah, I assumed the output after `dagger init` was correct. I'm talking about the pipeline, potentially simplifying it and avoiding cache invalidations. In this specific instance, you have a bunch of excludes when adding `m.SDKSourceDir`, but that directory comes from what you include in `ci/build/sdk.go` so it should already be "clean". As for the multi-stage, see how it's being done in python:

https://github.com/dagger/dagger/blob/379e812dee7c36713c30040324b99559bd74cf36/sdk/python/runtime/main.go#L268-L280

So those steps are separate from the main pipeline, which helps avoid cache invalidation because all of this produces a single input which is the generated/codegened file. Even if this sub-pipeline invalidates due to some temporary file during codegen installation for example, as long as the output is the same, it won't invalidate the main pipeline.

Just a suggestion, not a requirement 🙂

Reply:
> Or I should rename directory to `sdk` to avoid confusion?

The default right now is for `dagger init`to put the sources in a `dagger` sub-directory. So this puts the SDK in a `dagger/dagger` directory. On the other hand, I've add trouble filtering out `sdk` from the root of our repo because we have an `sdk` directory there. Maybe `dagger_sdk`? 🤷‍♂️

Reply:
> Yeah, I assumed the output after `dagger init` was correct. I'm talking about the pipeline, potentially simplifying it and avoiding cache invalidations. In this specific instance, you have a bunch of excludes when adding `m.SDKSourceDir`, but that directory comes from what you include in `ci/build/sdk.go` so it should already be "clean". As for the multi-stage, see how it's being done in python:
> 
> https://github.com/dagger/dagger/blob/379e812dee7c36713c30040324b99559bd74cf36/sdk/python/runtime/main.go#L268-L280
> 
> So those steps are separate from the main pipeline, which helps avoid cache invalidation because all of this produces a single input which is the generated/codegened file. Even if this sub-pipeline invalidates due to some temporary file during codegen installation for example, as long as the output is the same, it won't invalidate the main pipeline.
> 
> Just a suggestion, not a requirement 🙂

I see. I forgot that I just do filtering during packing sdk in ci/sdk/build.go. So we could delete this exclusion here to make it single source of truth for the exclusion list. 👍 

Back to the pipeline, I think I get it now after read the python source but need to play with it some more.

Reply:
> > Or I should rename directory to `sdk` to avoid confusion?
> 
> The default right now is for `dagger init`to put the sources in a `dagger` sub-directory. So this puts the SDK in a `dagger/dagger` directory. On the other hand, I've add trouble filtering out `sdk` from the root of our repo because we have an `sdk` directory there. Maybe `dagger_sdk`? 🤷‍♂️

I think that might be a good name for it. 👍

Reply:
Improved in https://github.com/dagger/dagger/pull/6967/commits/b56617b7f36897fb773bdec543f69a645e677be1 and rename the sdk to `dagger_sdk` in the latest changes. 🙇